x-pack/plugins/cloud_security_posture/public/pages/findings/use_findings_by_resource.ts

```diff
@@ -68,31 +82,102 @@ export const getFindingsByResourceAggQuery = ({
   },
 });
 
-export const useFindingsByResource = ({ enabled, index, query }: UseFindingsByResourceOptions) => {
+export const useFindingsByResource = ({
+  enabled,
+  index,
+  query,
+  size,
+  after,
+}: UseFindingsByResourceOptions) => {
   const {
     data,
     notifications: { toasts },
   } = useKibana().services;
 
-  return useQuery(
-    ['csp_findings_resource', { index, query }],
+  const [hasNextPage, setHasNextPage] = useState<boolean>(false);
+  const [afterKeys, setAfterKeys] = useState<FindingsByResourceAggregationsKeys[]>([]);
+
+  const queryResult = useQuery(
+    ['csp_findings_resource', { index, query, size, after }],
     () =>
       lastValueFrom(
         data.search.search<FindingsAggRequest, FindingsAggResponse>({
-          params: getFindingsByResourceAggQuery({ index, query }),
+          params: getFindingsByResourceAggQuery({ index, query, size, after }),
         })
       ),
     {
+      keepPreviousData: true,
+      refetchOnWindowFocus: false,
       enabled,
       select: ({ rawResponse }) => ({
+        after: rawResponse.aggregations?.groupBy?.after_key as
+          | FindingsByResourceAggregationsKeys
+          | undefined,
+        total: rawResponse.hits.total as number,
         page: rawResponse.aggregations?.groupBy.buckets.map(createFindingsByResource) || [],
       }),
       onError: (err) => showErrorToast(toasts, err),
     }
   );
+
+  const getNextKey = () => queryResult.data?.after;
+
+  const getPreviousKey = () => {
+    // when length is 0, we're on the 1st page, so no previous key
+    // when length is 1, we're on the 2nd page, so previous key is 'undefined'
+    if (afterKeys.length < 2) return undefined;
+
+    // We need the 'after_key' from the page that points to the previous page
+    return afterKeys[afterKeys.length - 2];
+  };
+
+  /**
+   * Keeps track of pages we can go back to
+   */
+  useEffect(() => {
+    if (!enabled) return;
+    setAfterKeys((keys) => {
+      if (!after) return [];
+      const hasKeyBeenSeen = keys.some((key) => isEqual(key, after));
+
+      return hasKeyBeenSeen
+        ? keys.slice(0, -1) // We go back one at a time
+        : [...keys, after];
+    });
+  }, [enabled, after]);
+
+  /**
+   * Keeps track of whether we have a next page
+   * checks if the next response has an `after_key`
+   * if it doesn't, then we already have the last page with data, and we disable the 'next' button
+   * NOTE: this is wasteful, but makes a better UX
+   */
+  useEffect(() => {
+    if (!enabled) return;
+    lastValueFrom(
+      data.search.search<FindingsAggRequest, FindingsAggResponse>({
+        params: getFindingsByResourceAggQuery({
+          index,
+          query,
+          size,
+          after: queryResult.data?.after,
+        }),
+      })
+    ).then((nextResult) =>
+      setHasNextPage(!!nextResult?.rawResponse?.aggregations?.groupBy?.after_key)
```

Comment:
this is bad, i don't like it at all 😄 

it's here because when we use `after` to paginate through results, the last result with data still has an `after_key` which will lead to an empty page. so if we let the user click "next" it will result in showing no results, as the response will have an empty `buckets` array. 

there's some discussion on this in slack, seems like a known issue. we can discuss this further IRL if needed.